File: Line Flows and Losses/python/lineflow_loss.py
"""
This program calculates lineflows and line losses of a power system by taking line voltages, currents,
and Bus Admittance Matrix as input from the user.
"""
import numpy as np
from tabulate import tabulate

"""
This function takes a complex number as input from the user

Args:
    prompt (str): The prompt to be displayed to the user

Returns:
    c(complex): The complex number entered by the user
"""
def get_complex_input(prompt):
    c = complex(input(prompt))
    return c


"""
This is the main function

Args:
    None

Returns:
    None
"""
def main():
    print("Enter the number of buses:")
    n = int(input())

    # Initialise matrices
    V = np.zeros((n, n), dtype=complex)
    I = np.zeros((n, n), dtype=complex)
    y = np.zeros((n, n), dtype=complex)

    [V,I,y] = get_input(n,V,I,y)
    [S,SL] = calculate_lineflow_loss(n,V,I,y)
    display_output(n,V,I,S,SL)


"""
This function takes the input from the user

Args:
    n (int): The number of buses
    V (numpy.ndarray): The voltage matrix
    I (numpy.ndarray): The current matrix
    y (numpy.ndarray): The admittance matrix

Returns:
    [V,I,y]: The voltage, current and admittance matrices

"""
def get_input(n,V,I,y):
    # Take voltages and currents as input
    for i in range(n):
        V[i, i] = get_complex_input(f"Enter the voltage at bus {i+1}: ")
        I[i, i] = get_complex_input(f"Enter the current at bus {i+1}: ")

    # Get the choice of input type (impedance or admittance)
    choice = int(input("Enter 1 for impedance and 2 for admittance: "))

    if choice == 1:
        # Calculate Admittance
        for i in range(n):
            for j in range(n):
                yij = get_complex_input(f"Enter the impedance between bus {i+1} and {j+1}: ")
                y[i, j] = 1 / yij
    elif choice == 2:
        # Take Admittance as input
        for i in range(n):
            for j in range(n):
                y[i, j] = get_complex_input(f"Enter the admittance between bus {i+1} and {j+1}: ")

    return [V,I,y]


"""
This function calculates the lineflows and line losses

Args:
    n (int): The number of buses
    V (numpy.ndarray): The voltage matrix
    I (numpy.ndarray): The current matrix
    y (numpy.ndarray): The admittance matrix

Returns:
    [S,SL]: The lineflows and line losses matrices
"""
def calculate_lineflow_loss(n,V,I,y):
    S = np.zeros((n, n), dtype=complex)
    SL = np.zeros((n, n), dtype=complex)
    # Calculate voltages
    for i in range(n):
        for j in range(n):
            if i != j:
                V[i, j] = V[i, i] - V[j, j]
                V[j, i] = V[j, j] - V[i, i]

    # Calculate currents
    for i in range(n):
        for j in range(n):
            if i != j:
                I[i, j] = y[i, j] * (V[i, i] - V[j, j])
                I[j, i] = y[j, i] * (V[j, j] - V[i, i])

   

    # Calculate Lineflows and line losses
    for i in range(n):
        for j in range(n):
            # Line Flows
            S[i][j] = V[i][j]*np.conj(I[i][j])
            S[j][i] = V[j][i]*np.conj(I[j][i])

            # Line losses
            SL[i][j] = S[i][j] + S[j][i]

    return [S,SL]

"""
This function displays the output

Args:
    n (int): The number of buses
    V (numpy.ndarray): The voltage matrix
    I (numpy.ndarray): The current matrix
    S (numpy.ndarray): The lineflows matrix
    SL (numpy.ndarray): The line losses matrix

Returns:
    None
"""
def display_output(n,V,I,S,SL):
    # Prepare data for tabulate 
    data = [] 
    for i in range(n):
        for j in range(n):
            data.append([ f"{i + 1}-{j + 1}", f"{V[i, j].real:.4f} + {V[i, j].imag:.4f}j", f"{I[i, j].real:.4f} + {I[i, j].imag:.4f}j", f"{S[i, j].real:.4f} + {S[i, j].imag:.4f}j", f"{SL[i, j].real:.4f} + {SL[i, j].imag:.4f}j" ])
    print(tabulate(data, headers=["Bus Pair", "Voltage", "Current", "Line Flow", "Line Loss"]))


if __name__ == "__main__":
    main()

File: Line Flows and Losses/java/lineflow_loss.java
import java.util.*;

/**
 * Main class to calculate line flows and losses
 * 
 * Functions:
 * 1. get_input(n,V,I,Y): Takes input from the user
 * 2. calculate_lineflow_loss(n,V,I,y): Calculates line flows and losses
 * 3. display_output(n,V,I,S,SL): Displays the output
 * 4. main(): Main function to run the program
 */
public class lineflow_loss {
    /**
     * Scanner object to take input from the user
     */
    static Scanner sc = new Scanner(System.in);

    /**
     * Main function to run the program
     * @param args
     */
    public static void main(String[] args) {
        System.out.println("Enter the number of buses:");
        int n = sc.nextInt();
        sc.nextLine();
        // Initialise matrices
        Complex[][] V = new Complex[n][n];
        Complex[][] I = new Complex[n][n];
        Complex[][] y = new Complex[n][n];
        
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                V[i][j] = new Complex(0, 0);
                I[i][j] = new Complex(0, 0);
                y[i][j] = new Complex(0, 0);
            }
        }

        ArrayList<Complex[][]> input = get_input(n, V, I, y);
        V = input.get(0);
        I = input.get(1);
        y = input.get(2);

        ArrayList<Complex[][]> output = calculate_lineflow_loss(n, V, I, y);
        Complex[][] S = output.get(0);
        Complex[][] SL = output.get(1);

        display_output(n, V, I, S, SL);
    
    }

    /**
     * Function to take input from the user
     * @param n total no.of buses
     * @param V voltage matrix
     * @param I current matrix
     * @param y line admittance matrix
     * @return ArrayList of matrices V,I,y
     */
    public static ArrayList<Complex[][]> get_input(int n, Complex[][] V, Complex[][] I, Complex[][] y) {
        // Take voltages and currents as input
        for(int i=0;i<n;i++){
            System.out.println("Enter voltage at bus "+(i+1));
            String s = sc.nextLine();
            V[i][i] = Complex.fromString(s);
            System.out.println("Enter current at bus "+(i+1));
            String t = sc.nextLine();
            I[i][i] = Complex.fromString(t);
        }

        // Get the choice of input type (impedance or admittance)
        int choice;
        System.out.println("Enter 1 for impedance and 2 for admittance: ");
        choice = sc.nextInt();
        sc.nextLine();
        

        if (choice == 1) {
            // Take impedance as input and calculate Admittance
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    System.out.println("Enter the impedance between bus " + (i+1) + " and bus " + (j+1));
                    String s = sc.nextLine();
                    y[i][j] = Complex.fromString(s);
                    y[i][j] = y[i][j].reciprocal();
                }
            }
        } else if (choice == 2) {
            // Take Admittance as input
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    System.out.println("Enter the admittance between bus " + (i+1) + " and bus " + (j+1));
                    String s = sc.nextLine();
                    y[i][j] = Complex.fromString(s);
                }
            }
        }


        ArrayList<Complex[][]> res = new ArrayList<>();
        res.add(V);
        res.add(I);
        res.add(y);
        return res;
    }


    /**
     * Function to calculate line flows and losses
     * @param n total no.of buses
     * @param V voltage matrix
     * @param I current matrix
     * @param y line admittance matrix
     * @return ArrayList of matrices S,SL
     */
    public static ArrayList<Complex[][]> calculate_lineflow_loss(int n,Complex[][] V, Complex[][] I, Complex[][] y){
        Complex[][] S = new Complex[n][n];
        Complex[][] SL = new Complex[n][n];

        // Calculate voltages
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if(i!= j){
                    V[i][j] = V[i][i].subtract(V[j][j]);
                    V[j][i] = V[j][j].subtract(V[i][i]);
                }
            }
        }
        // Calculate currents
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (i != j) {
                    I[i][j] = y[i][j].multiply(V[i][i].subtract(V[j][j]));
                    I[j][i] = y[j][i].multiply(V[j][j].subtract(V[i][i]));
                }
            }
        }
        
        // Calculate Lineflows and line losses
        for(int i=0;i<n;i++){
            for(int j=0;j<n;j++){
                // Line Flows
                S[i][j] = V[i][j].multiply(I[i][j].conjugate());
                S[j][i] = V[j][i].multiply(I[j][i].conjugate());
                
                // Line losses
                SL[i][j] = S[i][j].add(S[j][i]);
            }
        }

        ArrayList<Complex[][]> res = new ArrayList<>();
        res.add(S);
        res.add(SL);
        return res;
    }


    /**
     * Function to display the output
     * @param n total no.of buses
     * @param V voltage matrix
     * @param I current matrix
     * @param S line flow matrix
     * @param SL line loss matrix
     */
    public static void display_output(int n, Complex[][] V,Complex[][] I, Complex[][] S,Complex[][] SL){
        // Display results (Optional, remove if you want no output at all)
        ArrayList<String[]> data = new ArrayList<>();  // Initialize an ArrayList to store data

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                String[] rowData = new String[5];
                rowData[0] = String.format("%d-%d", i + 1, j + 1);  // Bus Pair
                rowData[1] = String.format("%.4f + %.4fj", V[i][j].real, V[i][j].imaginary);  // Use %f instead of %.4f
                rowData[2] = String.format("%.4f + %.4fj", I[i][j].real, I[i][j].imaginary);  // Use %f instead of %.4f
                rowData[3] = String.format("%.4f + %.4fj", S[i][j].real, S[i][j].imaginary);  // Use %f instead of %.4f
                rowData[4] = String.format("%.4f + %.4fj", SL[i][j].real, SL[i][j].imaginary);  // Use %f instead of %.4f
                data.add(rowData);  // Add the row data to the list
            }
        }

        // Print the table header
        String[] headers = {"Bus Pair", "Voltage", "Current", "Line Flow", "Line Loss"};
        System.out.printf("%-10s %-20s %-20s %-20s %-20s%n",headers[0],headers[1],headers[2],headers[3],headers[4]);
        
        // Print each row of data
        for (String[] row : data) {
            System.out.printf("%-10s %-20s %-20s %-20s %-20s%n", row[0], row[1], row[2], row[3], row[4]);
        }
    }
    

    
}
